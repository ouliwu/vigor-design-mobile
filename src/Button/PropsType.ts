/**
 * Button properties
 */
export interface ButtonPropsType {
  /** 按钮样式 */
  type?: 'default' | 'primary' | 'warn';
  /** 是否禁用按钮 */
  disabled?: boolean;
  /** 按钮点触事件 */
  onPress?: (e?: React.SyntheticEvent) => void;
  /** 按钮是否显示加载中 */
  loading?: boolean;
  /** 额外的class名 */
  className?: string;
  /** 按钮active状态下额外的class名 */
  activeClassName?: string;
}
