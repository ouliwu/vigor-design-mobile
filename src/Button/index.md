---
title: Button
nav:
  title: 组件
  path: /components
---

## 基础使用

### type

通过 type 属性切换 Button 组件样式

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button>default</Button>
      <br />
      <Button type="primary">primary</Button>
      <br />
      <Button type="warn">warn</Button>
    </>
  );
};
```

### disabled

通过 disabled 属性设置按钮是否禁用

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button
        disabled={true}
        onPress={() => {
          console.log(1);
        }}
      >
        default
      </Button>
      <br />
      <Button disabled={true} type="primary">
        primary
      </Button>
      <br />
      <Button disabled={true} type="warn">
        warn
      </Button>
    </>
  );
};
```

<API></API>
