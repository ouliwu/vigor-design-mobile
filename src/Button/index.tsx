import React, { useState } from 'react';
import './index.scss';
import { ButtonPropsType } from './PropsType';
import Touchable from '@ant-design/mobile-touchable';
import classNames from 'classnames';

const prefixCls = 'h-button';

const Button: React.FC<ButtonPropsType> = props => {
  const {
    type,
    disabled,
    onPress,
    loading,
    className,
    activeClassName,
    children,
  } = props;
  const [cls, setCls] = useState({
    [prefixCls]: true,
    [`${prefixCls}-active`]: false,
    [activeClassName!]: false,
  });
  const innerCls = classNames({
    ...cls,
    [`${prefixCls}-${type}`]: true,
    [className!]: !!className,
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-loading`]: loading,
  });
  return (
    <Touchable
      disabled={loading || disabled}
      onPressIn={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: true,
          [activeClassName!]: true,
        });
      }}
      onPress={e => {
        onPress && onPress(e);
      }}
      onPressOut={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: false,
          [activeClassName!]: false,
        });
      }}
    >
      <div className={innerCls}>
        <div className={`${prefixCls}-loading-container`}>
          <div className={`${prefixCls}-loading-main`} />
        </div>
        <div className={`${prefixCls}-text-container`}>
          <span>{children}</span>
        </div>
      </div>
    </Touchable>
  );
};

Button.defaultProps = {
  type: 'default',
  disabled: false,
  loading: false,
  className: '',
  activeClassName: '',
};

export default Button;
