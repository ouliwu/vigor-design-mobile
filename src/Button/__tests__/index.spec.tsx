import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '..';

describe('Button', () => {
  it('Button render correctly', () => {
    const wrapper = shallow(<Button>default</Button>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('type', () => {
    const wrapper = shallow(<Button type="primary">primary</Button>);

    expect(wrapper.find('.h-button').hasClass('h-button-primary')).toBeTruthy();
  });
  it('onPress', () => {
    const onPress = jest.fn();
    const wrapper = shallow(<Button onPress={onPress}>button</Button>);

    wrapper.simulate('press');

    expect(onPress).toBeCalled();
  });
  it('disabled', () => {
    const wrapper = shallow(<Button disabled={true}>button</Button>);
    // react这边不能做到disabled禁用后测试对应事件不触发，
    // 因为enzyme底层并没有处理类似情况，
    // enzyme仅仅是获取判断时候存在对应handler，如果存在直接调用
    expect(
      wrapper.find('.h-button').hasClass('h-button-disabled'),
    ).toBeTruthy();
  });
  it('className', () => {
    const wrapper = shallow(<Button className="propsCls">button</Button>);

    expect(wrapper.find('.h-button').hasClass('propsCls')).toBeTruthy();
  });
  it('activeClassName', () => {
    const wrapper = shallow(<Button activeClassName="propsCls">button</Button>);

    wrapper.simulate('pressIn');
    expect(wrapper.find('.h-button').hasClass('propsCls')).toBeTruthy();

    wrapper.simulate('pressOut');
    expect(wrapper.find('.h-button').hasClass('propsCls')).toBeFalsy();
  });
});
