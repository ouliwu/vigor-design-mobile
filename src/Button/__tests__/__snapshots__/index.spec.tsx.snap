// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button render correctly 1`] = `
<Touchable
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
>
  <div
    className="h-button h-button-default"
  >
    <div
      className="h-button-loading-container"
    >
      <div
        className="h-button-loading-main"
      />
    </div>
    <div
      className="h-button-text-container"
    >
      <span>
        default
      </span>
    </div>
  </div>
</Touchable>
`;
